Tried to shorten all sections (#269)
ITU-DevOps2026/ITU-minitwit: report/sections/02-1-ci_cd.md

## CI/CD
<!--- A complete description and illustration of stages and tools included in the CI/CD pipelines, including deployment and release of your systems. -->

Our CI/CD pipelines use GitHub Actions, which was chosen because it seemed the simplest option since our codebase is hosted on GitHub. We've created several workflows that run based on different event triggers. As mentioned in [Design and Architecture](#design-and-architecture), we have a production environment and an identical test environment; both are used in our CI/CD pipelines.  

> Our CI pipeline is triggered on a pull request to main and includes code analysis and automated testing. If the tests pass, a new image of the MiniTwit application is pushed to DockerHub and deployed to the test environment. This allows us to manually verify that new changes from a PR integrates well into our existing application, lowering the likelihood of unexpected issues occuring when the changes are deployed to production. 

Reviewer: ```suggestion
Our CI pipeline is triggered on a pull request to main and includes code analysis and automated testing. If the analysis and tests pass, a new image of the MiniTwit application is pushed to DockerHub and deployed to the test environment. This allows us to manually verify that new changes from a PR integrates well into our existing application, lowering the likelihood of unexpected issues occuring when the changes are deployed to production. 
```

> Our CI pipeline is triggered on a pull request to main and includes code analysis and automated testing. If the tests pass, a new image of the MiniTwit application is pushed to DockerHub and deployed to the test environment. This allows us to manually verify that new changes from a PR integrates well into our existing application, lowering the likelihood of unexpected issues occuring when the changes are deployed to production. 

Reviewer 2: ```suggestion
Our CI pipeline is triggered on a pull request to main and includes code analysis and automated testing. If the tests pass, a new image of the MiniTwit application is pushed to DockerHub and deployed to the test environment. This allows us to manually verify that new changes from a PR integrates well into our existing application, lowering the likelihood of unexpected issues occurring during deployment to production. 
```

<!--- a Codacy Static Code Analysis is run to analyze code quality and security issues. At the same time, the following workflows are triggered: a SonarQube Cloud static code analysis, a CodeQL code analysis to scan for security vulnerabilities and a Trivy security scan for Docker image vulnerabilities. There is also a workflow which runs automated tests, builds and pushes a new image of the MiniTwit application to DockerHub, and if all tests pass, this image is deployed to our test environment. This allows us to manually verify that the new changes from the PR integrate well into our existing application, and lowers the likelihood of unexpected issues when these changes are deployed to production. --->

<!--- Our CD pipeline starts running when a push to main happens, e.g. when a PR is merged. This pipeline includes the same steps as our CI pipeline explained above, except that the final step after passing the automated tests is the new image of MiniTwit being deployed to the production environment instead of test environment. This means our system is continuously deployed to production whenever changes have been reviewed and pass several checks. -->

> Figure \ref{fig:CICD_pipeline}, depicts a simplified diagram of the CI/CD pipeline with the names of the stages shown. Individual steps executed in each workflow are not included here. 

Reviewer 2: ```suggestion
Figure \ref{fig:CICD_pipeline} depicts a simplified diagram of the CI/CD pipeline with the names of the stages shown. Individual steps executed in each workflow are not included here. 
```

![Simplified CI/CD pipeline with no individual workflow steps shown \label{fig:CICD_pipeline}](./images/overall_cicd_pipeline.png){width=70%}

> Figure \ref{fig:test_workflow} is a diagram of the automated testing workflow. This workflow is reused in the workflow for deployment to test environment and the one for deployment to production environment. 

Reviewer 2: ```suggestion
Figure \ref{fig:test_workflow} shows the automated testing workflow. This is reused in the workflows for deployment. 
```

![Workflow running automated tests \label{fig:test_workflow}](./images/test_pipeline.png){width=70%}

Figure \ref{fig:deploy_prod_workflow} illustrates the workflow that deploys to our production environment. As mentioned before, it uses the workflow for tests shown in figure \ref{fig:test_workflow}.

![Workflow testing and deploying to production \label{fig:deploy_prod_workflow}](./images/prod_cd_pipeline.png)

> The workflow handling deployment to our test environment is similar to Figure \ref{fig:deploy_prod_workflow}. A diagram of it is included in [Appendix for CI/CD workflows](#appendix-for-ci-cd-workflows). Diagrams of the workflows for CodeQL, SonarQube and Trivy are found in the same place.

Reviewer 2: ```suggestion
The workflow handling deployment to our test environment is similar to Figure \ref{fig:deploy_prod_workflow}. A diagram of it is included in [Appendix for CI/CD workflows](#appendix-for-ci-cd-workflows) together with diagrams of the workflows for CodeQL, SonarQube and Trivy.
```
